File: Easy/Arrange Consonants and Vowels.cpp
//{ Driver Code Starts
#include<bits/stdc++.h>
using namespace std;

struct Node
{
	char data;
	struct Node *next;
	
	Node(int x){
	    data = x;
	    next = NULL;
	}
	
};

void printlist(Node *head)
{
	if (head==NULL)return;
	while (head != NULL)
	{
		cout << head->data << " ";
		head = head->next;
	}
	cout << endl;
}

void append(struct Node** headRef, char data)
{
	struct Node* new_node = new Node(data);
	struct Node *last = *headRef;

	if (*headRef == NULL)
	{
	    *headRef = new_node;
	    return;
	}
	while (last->next != NULL)
		last = last->next;
	last->next = new_node;
	return;
}

// task is to complete this function
struct Node* arrange(Node *head);

int main()
{
    int T;
    cin>>T;
    while(T--){
        int n;
        char tmp;
        struct Node *head = NULL;
        cin>>n;
        while(n--){
            cin>>tmp;
            append(&head, tmp);
        }
        head = arrange(head);
        printlist(head);
    }
	return 0;
}


// } Driver Code Ends


/*
Structure of the node of the linked list is as
struct Node
{
	char data;
	struct Node *next;
	
	Node(int x)
	{
	    data = x;
	    next = NULL;
	}
	
};
*/
// task is to complete this function
// function should return head to the list after making 
// necessary arrangements
struct Node* arrange(Node *head)
{
   //Code here
    Node* a = head;
    
    vector<char> vocali;
    vector<char> consonanti;

    while(a != NULL)
    {
        char c = a->data;
        
        if (c == 'a' || 
            c == 'e' || 
            c == 'i' ||
            c == 'o' ||
            c == 'u')    vocali.push_back(c);
        else             consonanti.push_back(c);

        a = a->next;
    }

    Node* b = head;
    
    for (auto it : vocali)
    {
        b->data = it;
        b = b->next;
    }

    for (auto it : consonanti)
    {
        b->data = it;
        b = b->next;
    }

    return head;
}

File: Medium/Arrange Consonants and Vowels.py
#User function Template for python3

"""
# Node Class

class Node:
    def __init__(self, val):
        self.data = val
        self.next = None

"""

class Solution:
    #Function to reverse a linked list.
    def swap(self, first, second):
        a = first.next
        b = second.next
        
        first.next = b
        second.next = second.next.next
        
        b.next = a
    
    def arrangeCV(self, head):
        # Code here
        if head is None:
            return None
        
        copy = head

        wowels = {'a','e','i','o','u'}
        
        v = 0
        c = 0 
        
        while copy:
            if copy.data in wowels:
                v = 1
            else:
                c = 1
            
            if v == 1 and c == 1:
                break
            
            copy = copy.next
        
        if v == 0 or c == 0:
            return head
        
        if head.data not in wowels:
            copy = head
            
            while copy.next:
                if copy.next.data in wowels:
                    break
                else:
                    copy = copy.next
                    
            copy2 = copy.next
            copy.next = copy.next.next
            copy2.next = head
            
            head = copy2
            
        temp1 = head
        temp2 = head.next
        
        while temp1.next:
            if temp1.next.data in wowels:
                temp1 = temp1.next
            else:
                break
        
        while True:
            if temp2.data in wowels:
                temp2 = temp2.next
            else:
                break
        
        while temp1 and temp2.next:
            if temp2.next.data in wowels:
                self.swap(temp1, temp2)
                temp1 = temp1.next
            else:
                temp2 = temp2.next
        
        
        return head



#{ 
 # Driver Code Starts
# Node Class
class Node:

    def __init__(self, val):
        self.data = val
        self.next = None


# Linked List Class
class Linked_List:

    def __init__(self):
        self.head = None
        self.tail = None

    def insert(self, val):
        if self.head is None:
            self.head = Node(val)
            self.tail = self.head
        else:
            self.tail.next = Node(val)
            self.tail = self.tail.next


def printList(head):
    tmp = head
    while tmp:
        print(tmp.data, end=' ')
        tmp = tmp.next
    print()


if __name__ == '__main__':
    for i in range(int(input())):
        n = int(input())
        arr = [str(x) for x in input().split()]

        lis = Linked_List()
        for i in arr:
            lis.insert(i)

        newHead = Solution().arrangeCV(lis.head)
        printList(newHead)

# } Driver Code Ends